### PR TITLE
Doxygen/znai integration

### PR DESCRIPTION
**Pull request by godexsoft**
Ok, so this is heavily WIP stuff of course. 

To try this you will need a few things: 
- Add doxygen to your dev container (Dockerfile)
- Enable `DOCS` option in cmake (`-DDOCS=ON` or through VSCode CMake Cache editor UI)
- Build the `docs` target
- Install znai (`brew install testingisdocumenting/brew/znai` on mac. If you are on Linux - bad luck, you are on your own :rage4:)

The above steps should generate `build/xml` doxygen output.
You can now try znai like so `znai --preview --port 3333` from the `docs` folder. 3333 can be replaced with your favourite port number. You can then navigate to `127.0.0.1:3333` to see the generated website. You can tweak things inside `docs` folder and znai will reload automatically.

### CODE REVIEW

File: docs/reference-manual/api-reference.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+-----
+:include-doxygen-compound: fn::filter_t
+
+:include-doxygen-compound: fn::filter_t::apply
+
+-----
+## All the operator() overloads
+
+> What i'd really like to be able to do is to grab all members of some entity and then use them as input to some template. So that we can generate similarly looking blocks of doc for each function or whatever.
+> I could not find a way to do it yet in Znai.
```
Comment (MykolaGolubyev) [inline docs/reference-manual/api-reference.md:27]: This sounds like a good idea. Could you please elaborate?

Comment (godexsoft) [reply docs/reference-manual/api-reference.md:27]: Certainly 👍 

I'd like something that allows me to create a "fragment" template, let's say to render documentation for one function. Then create another template for the class itself that iterates all member functions and renders them using the first "fragment" template.

In Freemarker it could look something like this.

- function.ftl:
```
- ${signature}
...
```

- class.ftl:
```
### ${name}
Member functions:
<#memberFunctions as mf>
    <@znai-import template="function.ftl" data=mf />
</#items>
---
Member variables:
<#memberVariables as mv>
    <@znai-import template="var.ftl" data=mv />
</#items>
```

- namespace.ftl:
```
## ${namespace.name}
<#namespace.classes as class>
    <@znai-import template="class.ftl" data=class />
</#items>
```

For this to work we would probably need to support something like `:include-doxygen-data: key` which simply returns **data** that then can be used with the templates. 

Hopefully the idea is more clear with the examples above.
Is this something that is already possible and i missed it? I have not been able to understand how to do it using the [templates in Znai's docs](https://testingisdocumenting.org/znai/layout/templates).

To recap, what i want is the ability to generate most of the documentation from Doxygen comments, not write the documentation by hand and only use Znai's Doxygen integration to get things like snippets of doc from individual functions and classes. It's very impressive how beautiful the documentation can be made with Znai but at this point it looks like i'd have to essentially write ALL of the documentation by hand, effectively copying the structure of the codebase instead of generating from it. Thanks for looking into this 🙇

File: docs/file-name.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class JsClass {
+    constructor() {
+    }
+}
+
+export default JsClass
```
Comment (godexsoft) [inline docs/file-name.js:6]: @MykolaGolubyev it would also be nice to automatically add `\n` at the end of each generated file to satisfy github 🥇

File: docs/reference-manual/api-reference.md (hunk 1 above)
Comment (MykolaGolubyev) [reply docs/reference-manual/api-reference.md:27]: I like the idea to expand Znai templates to be more powerful. 
Template you are looking to build, how different they are going to be from Doxygen class compound (https://testingisdocumenting.org/znai/CPP/auto-reference#utilssecondanotherclass) ?

I do want to make default doxygen support to be viable so most people wouldn't need to generate their own template. 

Also I realize that Znai doesn't have everything you need right now and that can be annoying. Would love to use this opportunity to improve the doc system based on your feedback and features requests.

Comment (godexsoft) [reply docs/reference-manual/api-reference.md:27]: The compound stuff is actually not bad at all. I have tried to use those and the output is pretty decent. 

The problem i have is that i can't automate the creation of a list of all classes in the namespace `fn` for example. Znai sort-of forces me to write the compound statement for each class by hand. 

I understand that this may be the way Znai was designed but i feel like you could add a few things here and there and it will allow us to generate things instead of explicitly list them. Just think about us adding a new class to `fn` namespace.. i'd like our CI to regenerate the beautiful doc through Znai instead of having to manually add the new class to the docs.. even if it's just a "compound" under a new heading.

Btw., nothing is annoying! 💙 I can just see how Znai can be improved. If you want to make it happen - great and Thank you. If not, then we will just have to write those by hand 👍 
Either way, thanks for taking the time to comment here!

Comment (MykolaGolubyev) [reply docs/reference-manual/api-reference.md:27]: To provide more context, the idea behind manual specification of classes is so then you can move across files or group in a more meaningful way. Also separate some classes with diagrams or some other visuals. 

Wonder if adding `compound` for the namespace makes sense. But then it will generate a bunch of things in a single file...

Comment (godexsoft) [reply docs/reference-manual/api-reference.md:27]: Yes, i see how that could be useful and i love the idea that there are essentially unlimited possibilities in terms of visuals (a very impressive set of features indeed!). 

I also realized the same problem - if we want to generate everything for a namespace it would either have to be all in one, potentially giant, file (not great and against Znai's ideas) or Znai would need to somehow support generating parts of the TOC using Freemarker or something like that (seems a bit overcomplicated). 

I think we can try to use Znai as it is and see how it goes. I'd also like to get it to work as part of our CI on this repo.
If you want a few easy improvements meanwhile:
- Automatically add `\n` at the end of generated files 
- Support non-capitalized labels/titles in TOC
    - In our library we have monadic operations such as `and_then`, `filter`, etc. and I'd like to have a top-level file for each of them. The title of such file would be `and_then`. But in the TOC it appears as `And_then` (automatic capitalization I assume). I'd like to be able to turn it off. Is there already a way that i missed?

Comment (godexsoft) [reply docs/reference-manual/api-reference.md:27]: A few more observations while I did not forget:
- `noexcept` does not appear to be propagated in the signatures - this one is important i'd say
- same with attributes like `[[nodiscard]]` - this one might be ok, not a big deal, although i'd prefer to have those too
- adding a `@note` somewhere in the doxygen comment cuts the resulting documentation short (at least for `:include-doxygen-doc:`). i.e. the `@note` and anything behind it is not rendered

File: docs/file-name.js (hunk 2 above)
Comment (MykolaGolubyev) [reply docs/file-name.js:6]: will add. To clarify, this file is a part of auto scaffolded znai example project. You probably should not use it/commit it.

File: docs/reference-manual/api-reference.md (hunk 1 above)
Comment (MykolaGolubyev) [reply docs/reference-manual/api-reference.md]: These are good ones! Clearly our use case for C++ was way simpler. Adding GitHub issues and will prioritize:
https://github.com/testingisdocumenting/znai/issues/1180
https://github.com/testingisdocumenting/znai/issues/1181
https://github.com/testingisdocumenting/znai/issues/1182

Happy you tried the presentation mode! 
Do you have a suggestion about how to split into slides? I assume you want to use it to present and have a certain flow in mind

File: docs/snippets/filter/expected_error.cpp [hunk 3]
```diff
@@ -0,0 +1,8 @@
+struct Error { std::string what; };
+fn::expected<int, Error> ex{12};
+
+auto value = ex 
+    | filter([](auto &&i) { return i >= 42; }, 
+             [](auto) { return Error{"Less than 42"}; });
+
+REQUIRE(value.error().what == "Less than 42");
```
Comment (MykolaGolubyev) [inline docs/snippets/filter/expected_error.cpp:8]: I noticed you have all the small files for (I assume) demonstrating compiler behavior. 
Just in case, wanted to ask if you know that in other cases you can extract sub snippets from larger files using a variety of approaches.

File: docs/reference-manual/api-reference.md (hunk 1 above)
Comment (MykolaGolubyev) [reply docs/reference-manual/api-reference.md]: > I'd also like to get it to work as part of our CI on this repo
what is missing? 
The crud way to get znai "binary" using wget from maven central: https://repo.maven.apache.org/maven2/org/testingisdocumenting/znai/znai-dist/1.67/znai-dist-1.67-znai.zip

there is a bit better integration with Maven build tool, but that is Java.

Close to ideal would be to have a GitHub action, but I haven't created any action so have a limited exposure. If this is something that you think could be useful - I will try to investigate

File: docs/snippets/filter/expected_error.cpp (hunk 3 above)
Comment (godexsoft) [reply docs/snippets/filter/expected_error.cpp:8]: I have not tried to extract `@code` blocks from doxygen if that's what you are referring to. Does that work with Znai? What are the other ways?

Even if it's possible though, personally i'm not a fan of adding code examples into comments in the actual codebase - there is enough code in there as it is. We can adjust if others disagree but imo small snippet files are the better choice seeing that we are essentially writing most of the documentation by hand anyway (as opposed to generating the entire doc doxygen style).

File: docs/reference-manual/api-reference.md (hunk 1 above)
Comment (godexsoft) [reply docs/reference-manual/api-reference.md]: Maven is unfortunately not of much use for us in C++.. 

> Close to ideal would be to have a GitHub action, but I haven't created any action so have a limited exposure. If this is something that you think could be useful - I will try to investigate

GitHub action would probably be the best. I'm happy to try embed Znai into our dev container and use it from CMake for now though. Have not tried yet. Will let you know if i run into any issues 👍 

> Do you have a suggestion about how to split into slides? I assume you want to use it to present and have a certain flow in mind

Did not really think this through yet - just thought it's a very cool feature and decided to try it out. I can see how it could be useful when pitching the library to others.

Perhaps a natural-ish flow would be to display (slide after slide)
- Name of the class and the text part (description of the class)
- All signatures on one slide if there are not many; or 
- Signature after signature if there are many (including parameters)

It's kinda hard to define what is "many" in this case as it also depends on the amount of text and parameters.. perhaps you can compute using total count of characters to keep it simple.

File: docs/file-name.js (hunk 2 above)
Comment (godexsoft) [reply docs/file-name.js:6]: Thanks. I figured not all generated files will end up in the repo but i'd still appreciate if it by default made github happy 🥇

File: docs/doxygen.json [hunk 4]
```diff
@@ -1,3 +1,3 @@
 {
-  "indexPath": "../build/xml/index.xml"
+  "indexPath": "../.build/xml/index.xml"
```
Comment (godexsoft) [inline docs/doxygen.json:2]: This is problematic. We are hardcoding a build folder name here.
It could be worked around by symlink or something like that from CMake.

File: cmake/Docs.cmake [hunk 5]
```diff
@@ -0,0 +1,27 @@
+find_package(Doxygen REQUIRED)
```
Comment (MykolaGolubyev) [inline cmake/Docs.cmake:1]: oh wow! your whole pipeline is so clean! 
I am scared of docker containers, and never had to set up them myself. But your code sells them very well.

Comment (godexsoft) [reply cmake/Docs.cmake:1]: Thanks 👍 I think there is still room for improvement but by using docker we can indeed remove a lot of the configuration and env setup from the CI itself.

File: docs/snippets/filter/expected_error.cpp (hunk 3 above)
Comment (MykolaGolubyev) [reply docs/snippets/filter/expected_error.cpp:8]: Not a fan of adding code examples to comments as well. Was wondering if you saw this: https://testingisdocumenting.org/znai/snippets/snippets-manipulation

File: docs/reference-manual/api-reference.md (hunk 1 above)
Comment (MykolaGolubyev) [reply docs/reference-manual/api-reference.md]: This is helpful, thank you.  After my local fix `noexcept(true)` renders as `noexcept`, will this work? In generated doxygen XML I don't see a distinction between `noexcept(true)` and `noexcept`.

Re `[[nodiscard]]`, don't see anything related to it in my locally generated XML. Is there a chance you could take a look at your generated XML and check if there is any data related to it?

Comment (godexsoft) [reply docs/reference-manual/api-reference.md]: In my case i do see some stuff like: `<highlight class="normal"><sp/><sp/>[[nodiscard]]<sp/></highlight>` in the XML output.

As for `noexcept(true)` vs `noexcept` - they are equivalent. 
What would be nice to output though is if in `noexcept(CONDITION)` the `CONDITION` is not a simple `true`. 
Idk. how well that is supported by Doxygen though.

Comment (MykolaGolubyev) [reply docs/reference-manual/api-reference.md]: ```<highlight class="normal"><sp/><sp/>[[nodiscard]]<sp/></highlight>``` is probably for the entire file content. There is another XML file where the args are parsed and there are also things like `const="yes"`. E.g:
```
      <memberdef kind="function" id="functional_8hpp_1ad744a75285a42e1302c54835d7298b4a" prot="public" static="no" const="no" explicit="no" inline="no" noexcept="yes" virt="non-virtual">
        <type>void</type>
        <definition>void repro_case::test1</definition>
        <argsstring>(int par1, std::string const &amp;par2) noexcept</argsstring>
        <name>test1</name>
        <qualifiedname>repro_case::test1</qualifiedname>
        <param>
          <type>int</type>
          <declname>par1</declname>
        </param>
```
That's what Znai parses

Comment (godexsoft) [reply docs/reference-manual/api-reference.md]: Oh, in that case probably you just don't have the information to display attributes (e.g. `[[nodiscard]]`). To be honest this may be an oversight on Doxygen's side as I'm not really seeing these attributes in the HTML docs generated by Doxygen for another project either. As i said before - attributes would be _nice_ but are not crucial.

File: .github/workflows/docs.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+          cmake -DDOCS=True ..
+          cmake --build . --target export_docs
+      
+      - name: Setup Pages
+        uses: actions/configure-pages@v4
+
+      - name: Upload artifact
+        uses: actions/upload-pages-artifact@v3
+        with:
+          path: ./.build/docs
```
[the remark this answers is not in the document]
Comment (MykolaGolubyev) [reply .github/workflows/docs.yml:43]: Ah right. One way to "improve" it a bit is to use `docs` as `docId` (e.g. https://liftwizard.io/docs/configuration/json5-configuration)

But also wonder (haven't tried) if you pass an empty string or maybe forward slash as docId. 

The reason for docId internal requirement is so urls can be treated the same way /docs/my-page and /docs/my-page/ (extra slash). Otherwise relative linking (..) would be messed up. 

If you go with `docs` then you can imagine having `blog` or something like that. Not ideal trade off I understand. 

Let me know if empty docId doesn't work and I will check the code to make it work.

[the remark this answers is not in the document]
Comment (godexsoft) [reply .github/workflows/docs.yml:43]: Thanks for the suggestions. Trying now.
- If i specify empty (``) or a space (` `) for doc-id, then i get `Missing argument for option: doc-id`
- Specifying a forward slash (`/`) yields an exception: 
  ```
  generating the rest of HTML pages
  deploying index.md to /index.html
  Exception in thread "main" java.lang.RuntimeException: error during rendering of index: java.nio.file.AccessDeniedException: /index.html
  ```
Looks like this use-case is just not supported atm.
Perhaps it will be easy to fix. I like the forward slash idea. I feel like it's almost working except for the permissions 🤞

[the remark this answers is not in the document]
Comment (MykolaGolubyev) [reply .github/workflows/docs.yml:43]: https://github.com/testingisdocumenting/znai/pull/1188

[the remark this answers is not in the document]
Comment (godexsoft) [reply .github/workflows/docs.yml:43]: Hi @MykolaGolubyev, hope you are doing well :+1: 
I see that the issue got merged some time ago - any release plans for the coming future? :bow:

[the remark this answers is not in the document]
Comment (MykolaGolubyev) [reply .github/workflows/docs.yml:43]: Thanks for the reminder! Released 1.68
Let me know how it goes. I can make another release soon

[the remark this answers is not in the document]
Comment (godexsoft) [reply .github/workflows/docs.yml:43]: @MykolaGolubyev thanks for your help on this! 

Znai now generates to root correctly when i specify `--doc-id '""'`.
Note that it does not work without the extra single quotes as it errors out the same as before with "empty" otherwise.

However, now it appears to ignore [my index page and opens straight to the first item in TOC](http://libfn.org). Any quick way to disable this behaviour you know of?
I have not yet checked the other features you added since 1.85. Will keep you posted when I do. 👍

[the remark this answers is not in the document]
Comment (MykolaGolubyev) [reply .github/workflows/docs.yml:43]: Could you please try to remove doc-id parameter altogether?

[the remark this answers is not in the document]
Comment (MykolaGolubyev) [reply .github/workflows/docs.yml:43]: ah sorry, there is an automatic name generation. I am looking into why index automatically redirects.
Feel free to provide more feedback so I can pack them all together for the next release

[the remark this answers is not in the document]
Comment (godexsoft) [reply .github/workflows/docs.yml:43]: > Could you please try to remove doc-id parameter altogether?

This does not appear to work. It creates the `no-id-specified` folder on export then.

> Feel free to provide more feedback so I can pack them all together for the next release

Sorry, did not find the time to write more documentation yet. So i don't really have feedback outside of what we covered before in the discussion on this pr. 🙇

[the remark this answers is not in the document]
Comment (MykolaGolubyev) [reply .github/workflows/docs.yml:43]: The fix was released as 1.69

[the remark this answers is not in the document]
Comment (godexsoft) [reply .github/workflows/docs.yml:43]: I have switched to the new version and it seems to do the right thing now 👍 🙇

[the remark this answers is not in the document]
Comment (MykolaGolubyev) [reply .github/workflows/docs.yml:43]: Let me know if you have docs deployed somewhere - I could take a look and maybe suggest a thing or two if you want. 
If you have some energy left, could you please post something here - https://github.com/testingisdocumenting/znai/discussions/824?sort=new

[the remark this answers is not in the document]
Comment (godexsoft) [reply .github/workflows/docs.yml:43]: @MykolaGolubyev latest docs are deployed to [libfn.org](https://libfn.org/) but we are yet to actually write the contents :rage4:

File: docs/reference-manual/pack.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
+---
+title: fn::pack
+---
+
+##### Defined in {style: "api", badge: "#include <functional/pack.hpp>"}
+
+---
+
+<!--
```
Comment (godexsoft) [inline docs/reference-manual/pack.md:9]: @MykolaGolubyev 👋 
I tried to add some more stubs today and ran into this error.
Not sure what's going on here. Hope you can shed some light on this.

FWIW i can see `fn::pack` in the generated doxygen XML. Any pointers are welcome 🙇

File: docs/meta.json [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
+  "title": "Functional",
+  "type": "Library",
+  "support": {
```
Comment (godexsoft) [inline docs/meta.json:4]: @MykolaGolubyev originally i had `viewOn` here and i'd keep it if there is a way to disable it on some pages. For example, i don't want this link on the landing `index.md` page but i'd like to have it for all the files inside of `reference-manual`. Is there a way to do this? I could not find it. Thanks

File: docs/reference-manual/pack.md (hunk 7 above)
Comment (MykolaGolubyev) [reply docs/reference-manual/pack.md:9]: what's the error message?

File: docs/meta.json (hunk 8 above)
Comment (MykolaGolubyev) [reply docs/meta.json:4]: just in case - viewOn should point to original md file that is the source for the page. 
What makes you want to hide it for some pages?

Comment (godexsoft) [reply docs/meta.json:4]: It generates urls automatically.. for the index.md it generates `url//index.` instead of `url/index.md`. If all urls would be correct, maybe i'd keep `viewOn` as is

File: docs/reference-manual/pack.md (hunk 7 above)
Comment (godexsoft) [reply docs/reference-manual/pack.md:9]: Copying the error out for visibility:

```
Exception in thread "main" java.lang.RuntimeException: 
markup parsing error:
    TOC item: reference-manual/pack
    full path: /Users/godexsoft/Development/functional/docs/reference-manual/pack.md
error handling include plugin <doxygen-doc>
  free param: fn::pack
  opts: {}
expected to find element <declname>
        at org.testingisdocumenting.znai.website.WebSite.throwParsingErrorMessage(WebSite.java:570)
        at org.testingisdocumenting.znai.website.WebSite.parseMarkupAndUpdateTocItemAndSearch(WebSite.java:562)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
        at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1117)
        at org.testingisdocumenting.znai.website.WebSite.parseMarkups(WebSite.java:501)
        at org.testingisdocumenting.znai.website.WebSite.parse(WebSite.java:193)
        at org.testingisdocumenting.znai.website.WebSite.parseAndDeploy(WebSite.java:179)
        at org.testingisdocumenting.znai.website.WebSite$Configuration.deployTo(WebSite.java:1034)
        at org.testingisdocumenting.znai.cli.ZnaiCliApp.generateDocs(ZnaiCliApp.java:197)
        at org.testingisdocumenting.znai.cli.ZnaiCliApp.start(ZnaiCliApp.java:99)
        at org.testingisdocumenting.znai.cli.ZnaiCliApp.start(ZnaiCliApp.java:70)
        at org.testingisdocumenting.znai.cli.ZnaiCliApp.main(ZnaiCliApp.java:75)
Caused by: java.lang.RuntimeException: error handling include plugin <doxygen-doc>
  free param: fn::pack
  opts: {}
expected to find element <declname>
        at org.testingisdocumenting.znai.parser.commonmark.MarkdownVisitor.handleIncludePlugin(MarkdownVisitor.java:232)
        at org.testingisdocumenting.znai.parser.commonmark.MarkdownVisitor.visit(MarkdownVisitor.java:155)
        at znaishaded.org.commonmark.node.CustomBlock.accept(CustomBlock.java:7)
        at znaishaded.org.commonmark.node.AbstractVisitor.visitChildren(AbstractVisitor.java:137)
        at znaishaded.org.commonmark.node.AbstractVisitor.visit(AbstractVisitor.java:28)
        at znaishaded.org.commonmark.node.Document.accept(Document.java:7)
        at org.testingisdocumenting.znai.parser.commonmark.MarkdownParser.parsePartial(MarkdownParser.java:98)
        at org.testingisdocumenting.znai.parser.commonmark.MarkdownParser.parse(MarkdownParser.java:66)
        at org.testingisdocumenting.znai.website.WebSite.parseMarkupAndUpdateTocItemAndSearch(WebSite.java:549)
        ... 10 more
Caused by: java.lang.IllegalArgumentException: expected to find element <declname>
        at org.testingisdocumenting.znai.utils.XmlUtils.lambda$nextLevelNodeByName$1(XmlUtils.java:87)
        at java.base/java.util.Optional.orElseThrow(Optional.java:403)
        at org.testingisdocumenting.znai.utils.XmlUtils.nextLevelNodeByName(XmlUtils.java:87)
        at org.testingisdocumenting.znai.doxygen.parser.DoxygenMemberParser.lambda$parseMember$0(DoxygenMemberParser.java:63)
        at java.base/java.util.stream.SpinedBuffer$1Splitr.forEachRemaining(SpinedBuffer.java:364)
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:782)
        at org.testingisdocumenting.znai.doxygen.parser.DoxygenMemberParser.parseMember(DoxygenMemberParser.java:62)
        at org.testingisdocumenting.znai.doxygen.parser.DoxygenMemberParser.parseXml(DoxygenMemberParser.java:44)
        at org.testingisdocumenting.znai.doxygen.parser.DoxygenMemberParser.parse(DoxygenMemberParser.java:38)
        at org.testingisdocumenting.znai.doxygen.parser.DoxygenCompoundParser.parseMember(DoxygenCompoundParser.java:66)
        at java.base/java.util.stream.SpinedBuffer$1Splitr.forEachRemaining(SpinedBuffer.java:356)
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:782)
        at org.testingisdocumenting.znai.doxygen.parser.DoxygenCompoundParser.parseXml(DoxygenCompoundParser.java:62)
        at org.testingisdocumenting.znai.doxygen.parser.DoxygenCompoundParser.parse(DoxygenCompoundParser.java:39)
        at org.testingisdocumenting.znai.doxygen.Doxygen.findAndParseCompound(Doxygen.java:136)
        at org.testingisdocumenting.znai.doxygen.Doxygen.getCachedOrFindAndParseCompound(Doxygen.java:101)
        at org.testingisdocumenting.znai.doxygen.plugin.DoxygenDocIncludePlugin.process(DoxygenDocIncludePlugin.java:58)
        at org.testingisdocumenting.znai.extensions.TrackingIncludePlugin.process(TrackingIncludePlugin.java:65)
        at org.testingisdocumenting.znai.parser.commonmark.MarkdownVisitor.handleIncludePlugin(MarkdownVisitor.java:228)
        ... 18 more
```

File: docs/meta.json (hunk 8 above)
Comment (MykolaGolubyev) [reply docs/meta.json:4]: this looks like a bug! could you please submit another issue?

Comment (godexsoft) [reply docs/meta.json:4]: @MykolaGolubyev all ready for you: https://github.com/testingisdocumenting/znai/issues/1210

File: docs/reference-manual/pack.md (hunk 7 above)
Comment (MykolaGolubyev) [reply docs/reference-manual/pack.md:9]: I was able to reproduce and believe the problem is that znai expects `declname` for every argument:
```
        <param>
          <type>std::in_place_type_t&lt; T &gt;</type>
        </param>
        <param>
          <type>auto &amp;&amp;...</type>
          <declname>args</declname>
        </param>
```
As a temporary workaround (if possible) a to change signature so that Doxygen generates `declname`. 
Longer term, do you have a preference what to use as a default?

Comment (godexsoft) [reply docs/reference-manual/pack.md:9]: I see. I'd go with empty string for unnamed params. I'm not sure how doxygen itself handles it though - i imagine we can't really doxy-comment on unnamed params. If some name is absolutely a must for znai then arg1, argN and args would do. or param1, paramN and params. Thats just my pref, i guess everyone will have their own

Comment (godexsoft) [reply docs/reference-manual/pack.md:9]: Any progress with this issue @MykolaGolubyev ?
The tmp workaround of course works but i don't think anyone would want to sacrifice code for docs so better support for unnamed arguments is very much desired! 🙇‍♂️

Comment (MykolaGolubyev) [reply docs/reference-manual/pack.md:9]: Thanks for reminding. Will allocate some time this week.

Comment (MykolaGolubyev) [reply docs/reference-manual/pack.md:9]: https://github.com/testingisdocumenting/znai/pull/1212

Comment (MykolaGolubyev) [reply docs/reference-manual/pack.md:9]: 1.70 is released

Comment (godexsoft) [reply docs/reference-manual/pack.md:9]: Thanks for letting me know 🙇 Will try as soon as i can.

Comment (MykolaGolubyev) [reply docs/reference-manual/pack.md:9]: 1.70 was released

File: docs/snippets/and_then/choice_value.cpp [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  CHECK(parse("null") == fn::choice(nullptr));
+  CHECK(parse("") == fn::choice(nullptr));
+  CHECK(parse("true") == fn::choice(true));
+  CHECK(parse("false") == fn::choice(false));
+  CHECK(parse("1025") == fn::choice(1025l));
+  CHECK(parse("10.25") == fn::choice(10.25));
+  CHECK(parse("2e9") == fn::choice(2e9));
+  CHECK(parse("5e9") == fn::choice(5e9));
+  CHECK(parse("foo").has_value(std::in_place_type<std::nullopt_t>));
+
```
Comment (Bronek) [inline docs/snippets/and_then/choice_value.cpp:48]: Please do not copy code excerpts like this into documentation; they can easily become obsolete. Instead do try to refer to existing code in unit tests - `tests/examples` is made for this. Ideally if Znai supported some form of code referral syntax, so you just put some kind of a comment inside e.g. `tests/examples/simple.cpp`. It's fine (a good practice even) to add more files to `tests/examples` specifically for this purpose.

File: docs/toc [hunk 10]
```diff
@@ -1,14 +1,31 @@
...
+    value_or
+choice
+    index
+    and_then
+    transform
+pack
+    index
+sum
+    index
+
```
Comment (Bronek) [inline docs/toc:31]: I like this structure (might not like it later in the life of the project, but today it's great)

**Review by Bronek — COMMENTED** [review]
Looking good, just please avoid copying code into documentation.

File: docs/toc (hunk 10 above)
Comment (godexsoft) [reply docs/toc:31]: Cool, you can always adjust it later (or i will on your request).

**Review by Bronek — APPROVED** [review]
